### PR TITLE
request 7 days of history

### PR DESCRIPTION
**Pull request by yenda**
High priority because should be included in nightlies asap so that people don't use messages over the weekend

### Summary:

- fetch 7 days of history when joining a chat
- make 7 24h requests to request 7 days because mailservers
ignores requests for a timespan > 24h
- make requests sequentially to avoid timeouts
- change mailserver after 3 timeouts on a request

### Steps to test:

- test that when recovering an account, 7 days of history is fetched
- check that when creating an account 7 days of history is fetched on joined channels
- test behavior with an upgraded app

status: ready

### CODE REVIEW

**Comment by ghost** [comment]
### Pull Request Checklist
- [x] Have you updated the documentation, if impacted (e.g. [docs.status.im](https://docs.status.im/docs/build_status.html))?

**Comment by chadyj** [comment]
Nice one Eric!

**Review by janherich — APPROVED** [review]

**Comment by chadyj** [comment]
Hey did you test this when Joining the Chinese chat? (old unicode channel `status 中文`, not new one `status-chinese`).  7 days of messages will be thousands.

File: src/status_im/transport/inbox.cljs [hunk 1]
```diff
@@ -305,34 +327,23 @@
...
-
-(fx/defn add-request
-  [{:keys [db] :as cofx} {:keys [topic request-id from to]}]
-  (log/info "offline inbox: message request " request-id "sent for inbox topic" topic "from" from "to" to)
-  {:db (-> db
-           (assoc-in [:transport.inbox/requests request-id] {:from from
-                                                             :to to
-                                                             :topic topic})
-           (assoc-in [:transport.inbox/topics topic :request-pending?] true))})
+  (if (= 3 (get-in [:transport.inbox/current-request :attemps] db))
```
Comment (rasom) [inline src/status_im/transport/inbox.cljs]: can we extract 3 to some constant?

File: src/status_im/transport/filters.cljs [hunk 2]
```diff
@@ -43,13 +43,24 @@
                     (re-frame/dispatch [:transport/messages-received js-error js-message]))]
      (add-filter! web3 params callback :discovery-topic))))
 
+(defn all-filters-added?
+  [{:keys [db]}]
+  (let [filters (into #{} (keys (get db :transport/filters)))
```
Comment (rasom) [inline src/status_im/transport/filters.cljs]: might be just `set`, not a big deal

File: src/status_im/transport/inbox.cljs [hunk 3]
```diff
@@ -210,18 +194,47 @@
...
+  of 24 hours, so we split requests per 24h spans if the last request was
+  done more than 24h ago"
+  [now-in-s [topic {:keys [last-request]}]]
+  (let [days        (conj
+                     (into [] (range (max last-request
+                                          (- now-in-s seven-days))
+                                     now-in-s
+                                     one-day))
+                     now-in-s)
+        day-ranges  (map vector days (drop 1 days))]
```
Comment (rasom) [inline src/status_im/transport/inbox.cljs]: `(rest days)`?

File: src/status_im/transport/inbox.cljs [hunk 4]
```diff
@@ -305,34 +328,24 @@
...
-
-(fx/defn add-request
-  [{:keys [db] :as cofx} {:keys [topic request-id from to]}]
-  (log/info "offline inbox: message request " request-id "sent for inbox topic" topic "from" from "to" to)
-  {:db (-> db
-           (assoc-in [:transport.inbox/requests request-id] {:from from
-                                                             :to to
-                                                             :topic topic})
-           (assoc-in [:transport.inbox/topics topic :request-pending?] true))})
+  (if (= maximum-number-of-attemps
```
Comment (rasom) [inline src/status_im/transport/inbox.cljs]: that shouldn't be a problem, but can we make it `<=` for future generations :D

Comment (yenda) [reply src/status_im/transport/inbox.cljs]: why ?

File: src/status_im/ui/components/connectivity/view.cljs [hunk 5]
```diff
@@ -20,7 +20,9 @@
      [react/text {:style    styles/text
                   :on-press (when mailserver-error?
                               #(re-frame/dispatch [:inbox.ui/reconnect-mailserver-pressed]))}
-      (i18n/label label)]]))
+      (if fetching?
+        (str (i18n/label label) " (" fetching? " requests left)")
```
Comment (rasom) [inline src/status_im/ui/components/connectivity/view.cljs]: if we show this "requests left" part to the end user let's create another label like `"{{label}} ({{fetching?}} requests left)"`

**Review by rasom — APPROVED** [review]

File: src/status_im/transport/inbox.cljs (hunk 4 above)
Comment (rasom) [reply src/status_im/transport/inbox.cljs]: idk, someone will add accidentally add another `inc` for this value and condition will never catch "too many attempts"

**Comment by yenda** [comment]
@asemiankevich expect it to take some time, requesting chat history for 7 days takes around 20 sec per chat

**Comment by asemiankevich** [comment]
@yenda it is like this for ~15 minutes. I have joined status and status-desktop

image

**Comment by asemiankevich** [comment]
I never was able to receive history for status-desktop in this PR; the tooltip above is always shown, also messages only for today are shown (not all). It could depend on super overload of mailservers or whatever idk. Also it does not depend wifi / mobile network , no luck anywhere

**Comment by yenda** [comment]
@asemiankevich ok I identified a problem with bad connections that sometimes end up being offline and pending requests is never resent which ends up with the app being stucked
It should be fixed now

**Comment by asemiankevich** [comment]
@yenda last build shows an error:
image

```
00:06 I/ReactNativeJS: NAME: "TypeError"
00:06 I/ReactNativeJS: MESSAGE: null is not an object (evaluating 'r.shh')
00:06 I/ReactNativeJS: generate_sym_key_from_password@index.android.bundle:12:6464800
00:06 I/ReactNativeJS: G__33995_34005@index.android.bundle:12:6489960
00:06 I/ReactNativeJS: index.android.bundle:12:1597370
00:06 I/ReactNativeJS: invoke_interceptor_fn@index.android.bundle:12:1552611
00:06 I/ReactNativeJS: invoke_interceptors@index.android.bundle:12:1552912
00:06 I/ReactNativeJS: execute@index.android.bundle:12:1555282
00:06 I/ReactNativeJS: handle@index.android.bundle:12:1563774
00:06 I/ReactNativeJS: re_frame$router$IEventQueue$_process_1st_event_in_queue$arity$1@index.android.bundle:12:1590859
00:06 I/ReactNativeJS: re_frame$router$IEventQueue$_run_queue$arity$1@index.android.bundle:12:1575844
00:06 I/ReactNativeJS: index.android.bundle:12:1579024
00:06 I/ReactNativeJS: re_frame$router$IEventQueue$_fsm_trigger$arity$3@index.android.bundle:12:1583235
00:06 I/ReactNativeJS: index.android.bundle:12:1577146
00:06 I/ReactNativeJS: index.android.bundle:24:1801
00:06 I/ReactNativeJS: k@index.android.bundle:24:643
00:06 I/ReactNativeJS: q@index.android.bundle:24:1007
00:06 I/ReactNativeJS: callImmediates@index.android.bundle:24:3202
00:06 I/ReactNativeJS: value@index.android.bundle:19:2904
00:06 I/ReactNativeJS: index.android.bundle:19:1377
00:06 I/ReactNativeJS: value@index.android.bundle:19:2621
00:06 I/ReactNativeJS: value@index.android.bundle:19:1347
00:06 I/ReactNativeJS: [native code]
00:06 I/ReactNativeJS: value@index.android.bundle:19:1282
00:06 I/ReactNativeJS: [native code]
00:06 I/ReactNativeJS: ***
```

https://app.testfairy.com/projects/4803622-status/builds/8626863/sessions/4403884861/?accessToken=LPZkxRRC6ric0d41PEbD47gfF/o

<blockquote><div><strong><a href="https://app.testfairy.com/projects/4803622-status/builds/8626863/sessions/4403884861/?accessToken=LPZkxRRC6ric0d41PEbD47gfF/o">TestFairy: Status 0.9.29 (10407) @2018</a></strong></div></blockquote>

**Comment by asemiankevich** [comment]
regarding the PR itself:
1. i was able to fetch history in **#test-team** starting from **10** October 
2. i was able to fetch history in **#status** starting from **7** October
3. i was able to fetch history in **#status-desktop** starting from **6** October 

It takes approximately 2:45 to fetch history (https://app.testfairy.com/projects/4803622-status/builds/8626863/sessions/4403884964/?accessToken=VG4MKTQkQDNZ/MLN3lZ7gTvyQg0 see till 2:45)  for these 3 channels (0 requests remaining). This time included app launch and account creation + joining chats. I would say it is too much to see history from user POV but it is a point to discuss.

So:
- lets fix this error upon launch (it does not happen in latest nightly)
- lets estimate how fast should this fetching be
- lets discuss it to be in the release or not based on previous 2 points
- also there is a suggestion: when i am in a public chat and history if still being fetched and i am deleting this public chats - the tooltip for remaining requests is still shown until it finishes, that is confusing for user as well as not needed when I deleted chat so we may want to stop fetching upon chat deletion

**Comment by chadyj** [comment]
> It takes approximately 2:45 to fetch history for these 3 channels (0 requests remaining). This time included app launch and account creation + joining chats. I would say it is too much to see history from user POV but it is a point to discuss.

This is a lot, especially if the user isn't interested in history.

How about we fetch on demand?  By default we load X days or messages. When the user pulls down we fetch more. Use can keep pulling down until no more results.

Personally I think showing, say, 10 or 20 messages by default is enough for a user to get context. This could be from the last 2 minutes in a busy chat, or a few days in slower chats. If the user wants to see more they can keep scrolling.  WDYT?

**Comment by yenda** [comment]
@asemiankevich should be fixed in next build, I assume you had a connection status change during app loading, I added a guard for that

@chadyj I think that is premature optimization for status-react. The volume of data is ridiculously low for a week worth of messages, we should first aim at fixing how the mailserver sends the data. For now it might be long but users won't be joining 3 different chats right after creating their accounts so I don't think it is a big problem, and I prefer consistency and reliability over speed at that point.

**Comment by chadyj** [comment]
> For now it might be long but users won't be joining 3 different chats right after creating their accounts so I don't think it is a big problem, and I prefer consistency and reliability over speed at that point.

For a new user, sure. But what about a returning user following several chats, or worse, 20+ like many of us at Status. If the numbers above are any indication it could be 15 minutes just to download new messages after a short absence.  Could be an issue?

Also, to be on the safe side @lukaszfryc can you please look at battery performance for this PR?

**Comment by yenda** [comment]
@chadyj it's one request for 24h, if you join after a short period of time it will only make small request. Also once status-go allows requesting multiple topics is a single request it will be much faster for that scenario.

Regarding battery perf it can't be worse than now unless you connect like once a week. but that just the downside to have full history as opposed to just a day

**Comment by asemiankevich** [comment]
@yenda @chadyj so when joining 1 public chat after creating an account it takes ~2.10 to get the history. What is confusing is that the very first message appears in chat after ~1 min waiting (see https://app.testfairy.com/projects/4803590-status/builds/8627396/sessions/4403985901/?accessToken=5rCJsCpAjc1IYYQi8xYSPxjL1Tc)

I think we will get a lot of reports `no messages in public chat` if we release it. 

Also agree with @chadyj on upgrade scenario - we will never know how many chats were open before upgrade. I have checked the following scenario - opened 3 chats (test-team, status and status-desktop) on release build and upgraded.  Seems the fetching is stuck cause i dont see the request number decreasing (https://app.testfairy.com/projects/4803590-status/builds/8627396/sessions/4403987117/?accessToken=crdF30hmbsRK-CkzTellJfvIiX0)

Seems i won't see the history after upgrade (was waiting for ~10 min). 

<blockquote><div><strong><a href="https://app.testfairy.com/projects/4803590-status/builds/8627396/sessions/4403985901/?accessToken=5rCJsCpAjc1IYYQi8xYSPxjL1Tc">TestFairy: Status 0.9.29 (10417) @2018</a></strong></div></blockquote>
<blockquote><div><strong><a href="https://app.testfairy.com/projects/4803590-status/builds/8627396/sessions/4403987117/?accessToken=crdF30hmbsRK-CkzTellJfvIiX0">TestFairy: Status 0.9.29 (10417) @2018</a></strong></div></blockquote>

**Comment by asemiankevich** [comment]
@oskarth @cammellos please take a look. cc @annadanchenko

**Comment by chadyj** [comment]
@lukaszfryc would love to see how this PR impacts battery performance.

@yenda As a follow up to this PR what do you think about improving the "fetching" UI? https://github.com/status-im/status-react/issues/4345#issue-324458925

**Comment by lukaszfryc** [comment]
> @lukaszfryc would love to see how this PR impacts battery performance.

@chadyj Sorry, I missed that comment. Yes, it's good idea to test it when the PR is ready to merge. I will do it before the upcoming release. Probably today in the evening or tomorrow morning depending when the PR is finished

**Comment by asemiankevich** [comment]
I am able to get history for upgraded app. Timeframe is still ~ 2.10. Pelase check https://app.testfairy.com/projects/4803622-status/builds/8627852/sessions/4404015868/?accessToken=dUBhH0D/-FgRwkVVxi3z3zSTbuE

@lukaszfryc proceed with perf tests for latest build please

<blockquote><div><strong><a href="https://app.testfairy.com/projects/4803622-status/builds/8627852/sessions/4404015868/?accessToken=dUBhH0D/-FgRwkVVxi3z3zSTbuE">TestFairy: Status 0.9.29 (10522) @2018</a></strong></div></blockquote>

**Comment by yenda** [comment]
@asemiankevich is it worst that the current develop which doesn't even tell you how much requests are left ? I will make other PRs as soon as go team merges changes to improve the speed of requests but right now I don't think there is much we can do this time is the price to pay to have accurate history with no losses

**Comment by asemiankevich** [comment]
okay then lets merge it as is @yenda @chadyj

**Comment by lukaszfryc** [comment]
we can check the battery on develop. It's even better

**Comment by chadyj** [comment]
> we can check the battery on develop. It's even better

:-1:

@lukaszfryc I'd much prefer we spot battery regressions at PR rather than in release. Let's at least get some benchmarks so we can be confident moving forward.

**Comment by lukaszfryc** [comment]
@chadyj I will build nightly now and do the tests. Of course I'd like to to them as often and fast as possible, but they are quite time consuming.

**Comment by chadyj** [comment]
It is unfortunate this was merged without performance testing even though we established this could add several minutes of activity.

We are planning on cutting a release tomorrow for Hackathon and Devcon and we didn't do our due diligence on a big change.  Why would we want to do this?

cc @oskarth @annadanchenko

**Comment by yenda** [comment]
> It is unfortunate this was merged without performance testing even though we established this could add several minutes of activity.
> 
> We are planning on cutting a release tomorrow for Hackathon and Devcon and we didn't do our due diligence on a big change. Why would we want to do this?
> 
> cc @oskarth @annadanchenko

It's easy to switch back to only one day of history if fetching 7 days is too much of a drain, but not doing concurrent requests with a good chunk timing out and being processed multiple time (increasing both cpu and network activity) cannot be worst in terms of performance

**Comment by oskarth** [comment]
I agree with @chadyj. We had a lot of back and forth with battery perf, and this should be part of automated QA or QA. I wouldn't be against reverting this commit and doing it properly. The 24h window was chosen for a reason initially, so changing this should be tested properly (also, any reason we don't do this under a variable flag?).

It's probably not a good idea to merge when there's a lack of consensus in the the comment section.

@chadyj In future, I'd consider using that blocking option in review. Though it is meant for review of code, but we need a way to signal strongly that we disagree with just going ahead.

**Comment by oskarth** [comment]
@lukaszfryc what can we do to make it quicker?

I hope people still remember what happened with RN keyboard etc update around beta time. Let's avoid the same situation for latest release before devcon/hackathon...

---

If this change was introduced under a flag we could have this on nightly and leave release alone, for example

**Comment by lukaszfryc** [comment]
> @lukaszfryc what can we do to make it quicker?

There are couple of options. The best is to run E2E tests on real devices and monitor battery consumption OR CPU and memory consumption as a proxy for battery usage. @antdanchenko may have some timeline.

**Comment by lukaszfryc** [comment]
@oskarth @chadyj to be on the same page, I'm totally about testing battery, CPU and what not as soon as possible (PR level). Not having automated checks for it is our tech debt (we've just spent our time differently). But, I do not agree that doing it manually for **every PR** is the best use of our time. Having said that, if we feel strongly that some PR may introduce regressions in those areas I agree that we can freeze the PR and test it first instead of testing the develop before next release.

**Comment by chadyj** [comment]
Here is some personal experience using this change in the new Mac desktop nightly. These are first impressions from first 10 minutes.

- I got the dreaded "[We were not able to decrypt your data](https://github.com/status-im/status-react/issues/6112)" error when upgrading from today's earlier nightly
- Had to blow away the db and start again :poop:
- I recovered my account and 50 chats started appearing. 
- Fetching message for quite some time
- I joined some public channels like #status and #status-desktop
- The app is really unresponsive. 
- Typing is blocked. I start typing and there is nothing for a few seconds before characters appear
- Having problems scrolling
- Clicking around in the UI takes some time
- Doing something like sending a message took 5 seconds from hitting send until it appeared.
- CPU for status and ubuntu-server is ~50% and 70% even when idle (no fetching message displayed) 

Having older chats appear was good, but the performance appears like a regression.

**Comment by yenda** [comment]
@chadyj regarding the irresponsive UI we should open an issue on desktop for that, it is not related to this PR as the same effect can be achieved by scrolling up top the beginning of the history of a long chat.

on mobile this doesn't happen

**Comment by chadyj** [comment]
Ok @yenda trying this on ios:

- Like desktop above I also get a a "we detected a problem with the encryption key" error. Could not recover and had to delete the app. I've been updating nightlies for weeks every day and haven't had this issue until now. Not sure if this is just me, but if not, you are going to have a lot of pissed nightly users tomorrow.
- It showed "fetching messages" and downloaded messages from Oct 9 then stopped for about 30 seconds
- "fetching messages 7 requests" flashed a few more time over the next minute but would mainly stay on the screen. It stayed up for another 2 min without visibly downloading anything new. Only Oct 9 messages are displaying
- After another minute I saw a single chat at the top of the list from Oct 8.
- Joined a few public chats. See fetching messages 35 requests left. At this point it has been about 6 minutes and I only see 1-1's from Oct 9. Nothing newer.
- "fetching messages 35 requests" is still shown. Sometimes it flashes.
- The public chats I joined are empty, except 1 which has 3 messages.
- I took a look at Mailserver settings and I see it is jumping from mail server to mail server after 2-10 seconds
- After 10 minutes there aren't any new messages. Still see 35 requests left.  
- Gave up and quit the app. Logged back. Saw "fetching messages 35 requests" for a moment, then nothing. Still no new updates since Oct 9.  Mobile only picked up 15 chats, but Desktop picked up at 50.
- After 15 minutes the app doesn't seem to be doing anything. No more "fetching" messages and the mail server choice is staying put. Still see empty pub chats that should be full (eg #status) and no new 1-1 messages since Oct 9.

>  it is not related to this PR as the same effect can be achieved by scrolling up top the beginning of the history of a long chat.

I didn't have this issue on desktop before this build.

**Comment by asemiankevich** [comment]
@chadyj the encryption problem is hardly related to this PR https://github.com/status-im/status-react/issues/6112. 

Do you want @yenda to revert this commit or change the history back to 24 hours @chadyj ?

**Comment by lukaszfryc** [comment]
@asemiankevich did you notice the issues Chad mentioned when you tested the PR?

**Comment by lukaszfryc** [comment]
@asemiankevich @yenda I'd say that if issues @chadyj mentioned happen only on this PR, we should definitely revert it. It's not a release blocker but those issues probably are.

**Comment by cammellos** [comment]
@lukaszfryc before we revert, there were a few fixes that I believe got include in the latest nightly https://github.com/status-im/status-react/pull/6328 https://github.com/status-im/status-react/pull/6327 , could you please check if the issues are still present?

I'd rather keep reverting as a last option, moving forward if possible.

**Comment by rasom** [comment]
i would rather adjust it to fetch only 24h than revert the commit

**Comment by asemiankevich** [comment]
@rasom okay great, lets do that then.

**Comment by rasom** [comment]
https://github.com/status-im/status-react/pull/6333

**Comment by lukaszfryc** [comment]
> Like desktop above I also get a a "we detected a problem with the encryption key" error. Could not recover and had to delete the app. I've been updating nightlies for weeks every day and haven't had this issue until now. Not sure if this is just me, but if not, you are going to have a lot of pissed nightly users tomorrow.

@chadyj It's not related to this PR. Couple of people including me had issues with it yesterday, see https://github.com/status-im/status-react/issues/6112#issuecomment-429797224.

**Comment by chadyj** [comment]
> It's not related to this PR. Couple of people including me had issues with it yesterday

There were 9 commits in the nightly. What one causes the encryption key error? Should be pretty easy to track down?

https://ci.status.im/job/status-react/job/nightly/740/

<blockquote><div><strong><a href="https://ci.status.im/job/status-react/job/nightly/740/">status-react » nightly #740 [Jenkins]</a></strong></div></blockquote>

**Comment by lukaszfryc** [comment]
Btw, it's definitely a release blocker if this issue now appears also while upgrading from already released builds. I will test for it.

**Pull request merged**